### PR TITLE
JS: Add RegExpAlwaysMatches query

### PR DESCRIPTION
**Pull request by asgerf**
Adds a query to flag checks like `/[a-z]*/.test(x)` which always returns `true` since it can always match an empty substring at the beginning of `x`.

Find [a few results](https://lgtm.com/query/1482690140500333877).

Evaluation underway

### CODE REVIEW

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.qhelp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+<li>To match the whole input string, add anchors at the beginning and end of the regular expression.</li>
+<li>To search for an occurrence within the input string, consider what the shortest meaningful match is and restrict the
+regular expression accordingly, such as by changing a <code>*</code> to a <code>+</code>.</li>
+</ul>
+</p>
+
+</recommendation>
+<example>
+<p>
+In the following example, a regular expression is used to check the format of an string <code>id</code>.
```
Comment (max-schaefer) [inline javascript/ql/src/RegExp/RegExpAlwaysMatches.qhelp]:
```suggestion
In the following example, a regular expression is used to check the format of a string <code>id</code>.
```

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.ql [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/**
+ * @name Regular expression always matches
+ * @description Regular expression tests checks always find a match indicate dead code or a logic error
```
Comment (max-schaefer) [inline javascript/ql/src/RegExp/RegExpAlwaysMatches.ql]:
```suggestion
 * @description Regular expression tests that always find a match indicate dead code or a logic error
```

(perhaps)

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.ql [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  exists(RegExpTerm child | child = term.(RegExpStar).getAChild() |
+    child instanceof RegExpDot
+    or
+    child.(RegExpCharacterClass).isUniversalClass()
+  )
+}
+
+/**
+ * A call that searches for a regexp match within a string, but does not
+ * extract the capture groups or the matched string itself.
```
Comment (max-schaefer) [inline javascript/ql/src/RegExp/RegExpAlwaysMatches.ql]: Where is the "but does not (...)" bit checked?

File: javascript/ql/test/query-tests/RegExp/RegExpAlwaysMatches/tst.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+function optionalPrefix(x) {
+  return /^(https:)?/.test(x); // NOT OK
+}
+
+function mandatoryPrefix(x) {
+  return /^https:/.test(x); // NOT OK
```
Comment (max-schaefer) [inline javascript/ql/test/query-tests/RegExp/RegExpAlwaysMatches/tst.js]: Should this be "// OK"?

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.ql [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  not isUniversalRegExp(term) and
+  term = getEffectiveRoot(call.getRegExp()) and
+  (
+    call instanceof RegExpTestCall and
+    not isPossiblyAnchoredOnBothEnds(term) and
+    message = "This regular expression always matches when used in a test $@, as it can match an empty substring."
+    or
+    call instanceof RegExpSearchCall and
+    not term.getAChild*() instanceof RegExpDollar and
+    not term.getAChild*() instanceof RegExpSubPattern and
```
Comment (max-schaefer) [inline javascript/ql/src/RegExp/RegExpAlwaysMatches.ql]: This condition is slightly different from the other disjunct; could we have tests documenting (and perhaps even explaining) the difference?

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.ql (hunk 3 above)
Comment (asgerf) [reply javascript/ql/src/RegExp/RegExpAlwaysMatches.ql]: The doc comment attempts to explain why we don't include things like `exec` and `replace` calls. There isn't a corresponding check anywhere, we just don't include those calls.

Comment (max-schaefer) [reply javascript/ql/src/RegExp/RegExpAlwaysMatches.ql]: Right; I figured that out eventually and tried to remove my comment, but apparently unsuccessfully.

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.ql (hunk 5 above)
Comment (asgerf) [reply javascript/ql/src/RegExp/RegExpAlwaysMatches.ql]: added two tests

File: change-notes/1.24/analysis-javascript.md [hunk 6]
```diff
@@ -17,6 +17,7 @@
 | **Query**                                                                       | **Tags**                                                          | **Purpose**                                                                                                                                                                            |
 |---------------------------------------------------------------------------------|-------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | Cross-site scripting through exception (`js/xss-through-exception`) | security, external/cwe/cwe-079, external/cwe/cwe-116              | Highlights potential XSS vulnerabilities where an exception is written to the DOM. Results are not shown on LGTM by default. |
+| Regular expression always matches (`js/regex/always-matches`) | correctness, regular-expressions | Highlights regular expression checks that trivially succeed by matching an empty substring. |
```
Comment (mchammer01) [inline change-notes/1.24/analysis-javascript.md]: Sorry, forgot to comment on this. Are the results of this query shown on LGTM by default?

File: javascript/ql/test/query-tests/RegExp/RegExpAlwaysMatches/tst.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+function emptyAlt2(x) {
+  return /(^$|foo|bar)/.test(x); // OK
+}
+
+function emptyAlt3(x) {
+  return /((^$|foo|bar))/.test(x); // OK
+}
+
+function search(x) {
+  return /[a-z]*/.search(x); // NOT OK
```
Comment (max-schaefer) [inline javascript/ql/test/query-tests/RegExp/RegExpAlwaysMatches/tst.js]: This is marked as "NOT OK", but doesn't appear in the `.expected` file.

File: javascript/ql/test/query-tests/RegExp/RegExpAlwaysMatches/tst.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+function search2(x) {
+  return /[a-z]/.search(x); // OK
+}
+
+function lookahead(x) {
+  return /(?!x)/.search(x); // OK
+}
+
+function searchPrefix(x) {
+  return /^foo?/.search(x); // NOT OK - `foo?` does not affect the returned index
```
Comment (max-schaefer) [inline javascript/ql/test/query-tests/RegExp/RegExpAlwaysMatches/tst.js]: Ditto.

File: javascript/ql/src/RegExp/RegExpAlwaysMatches.qhelp [hunk 9]
```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+There are several built-in JavaScript functions that search for a regular expression match within a string,
+such as <code>RegExp.prototype.test</code> and <code>String.prototype.search</code>.
+If the regular expression is not anchored, the regular expression does not need to match the whole string;
```
[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/ql/src/RegExp/RegExpAlwaysMatches.qhelp]: Rephrased as bit, PTAL.

File: change-notes/1.24/analysis-javascript.md (hunk 6 above)
Comment (asgerf) [reply change-notes/1.24/analysis-javascript.md]: Yes, updated.

**Comment by asgerf** [comment]
[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/regexp-always-matches_1578385324629) looks fine. We get some FPs in the test suites of test262, node, etc.

**Pull request merged**